src/cmd/tools/doctest/doctest.go

```diff
@@ -0,0 +1,193 @@
+)
+
+func init() {
+	flag.StringVar(&file, "file", "", "File to be parsed")
+	flag.StringVar(&destDir, "dest", "", "Directory where the output will be saved")
+
+	flag.BoolVar(&markdown, "markdown", true, "generate a markdown file based on the source (default=true)")
+	flag.BoolVar(&bash, "bash", true, "generate a bash script based on the source (default=true)")
+}
+
+func main() {
+	flag.Parse()
+
+	// this is not ideal
+	if file == "" {
+		panic("'file' flag is required")
+	}
+
+	if destDir == "" {
+		destDir = filepath.Dir(file)
+	}
+
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Fatalf("unable to read source file: %v\n", err)
+	}
+
+	operationRegEx := regexp.MustCompile(operationRegEx)
+	validationRegEx := regexp.MustCompile(validationRegEx)
+
+	if bash {
+		if err := bashScript(f, destDir, file, operationRegEx, validationRegEx); err != nil {
+			log.Fatalf("unable to create bash script: %v", err)
+		}
+	}
+
+	if markdown {
+		if err := markdownFile(f, destDir, file, validationRegEx, operationRegEx); err != nil {
+			log.Fatalf("unable to create markdown file: %v", err)
+		}
+	}
+}
+
+func markdownFile(contents []byte, destDir, fileName string, validationRegEx *regexp.Regexp, opRegex *regexp.Regexp) error {
+	// create new markdown file based on name of source file
+	mdFileName := destDir + "/" + strings.Replace(filepath.Base(fileName), ".source", "", -1)
+	mdFile, err := os.Create(mdFileName)
+	if err != nil {
+		return err
+	}
+	defer mdFile.Close()
+	mdFile.Chmod(0644)
+
+	mdFile.Write([]byte("<!-- Code generated by doctest. DO NOT EDIT. -->\n"))
+	sourceFile := strings.TrimPrefix(fileName, goPath)
+	mdFile.Write([]byte(fmt.Sprintf("<!-- source: %s -->\n\n", sourceFile)))
+
+	// find all of the operations in the source file
+	matches := opRegex.FindAll(contents, 100)
+	for _, match := range matches {
+		validations := validationRegEx.FindAll(match, 100)
+		if len(validations) == 0 {
+			continue
+		}
+
+		// remove the validations and if the operation block is empty, replace it with nothing
+		noValidationBlock := bytes.Replace(match, validations[0], nil, -1)
+
+		if string(noValidationBlock) == fmt.Sprintf(operationOpen+"\n"+operationClose) {
+			contents = bytes.Replace(contents, match, []byte(""), -1)
+		}
+	}
+
+	// convert tags to markdown syntax
+	OpOpen := regexp.MustCompile(operationOpen)
+	OpClose := regexp.MustCompile(operationClose)
+	OpJSONOpen := regexp.MustCompile(operationJSONOpen)
+	OpJSONClose := regexp.MustCompile(operationJSONClose)
+
+	contents = OpOpen.ReplaceAll(contents, []byte("```"))
+	contents = OpClose.ReplaceAll(contents, []byte("```"))
+	contents = OpJSONOpen.ReplaceAll(contents, []byte("```json"))
+	contents = OpJSONClose.ReplaceAll(contents, []byte("```"))
+	contents = validationRegEx.ReplaceAll(contents, nil)
+	mdFile.Write(contents)
+
+	return nil
+}
+
+func bashScript(contents []byte, destDir, fileName string, operationRegEx, validationRegEx *regexp.Regexp) error {
```

Comment:
nit: take the destination filename, don't assume it's the same name

Reply:
I feel like it should just be the same name. would be kind of weird to have different names